scripts/object_storage_cleaner/storage_cleaner.py

```diff
@@ -0,0 +1,639 @@
+
+
+def parse_since_cutoff(value: str) -> datetime:
+    """
+    Parse a duration like '3s', '30d', '2w', '3 days' and return a datetime object.
+
+    Args:
+        value: The duration string to parse.
+
+    Returns:
+        A UTC datetime equal to now minus the parsed duration.
+    """
+    value = value.strip().lower()
+
+    pattern = (
+        r"^(\d+)\s*(second|seconds|minute|minutes|hour|hours|day|days|week|weeks)$"
+    )
+    match = re.match(pattern, value)
+    if not match:
+        raise argparse.ArgumentTypeError(
+            "Invalid duration. Use: '3 seconds', '10 hours', '15 minutes', '30 days', '2 weeks'"
+        )
+
+    amount = int(match.group(1))
+    unit = match.group(2)
+
+    if unit in ("second", "seconds"):
+        delta = timedelta(seconds=amount)
+    elif unit in ("minute", "minutes"):
+        delta = timedelta(minutes=amount)
+    elif unit in ("hour", "hours"):
+        delta = timedelta(hours=amount)
+    elif unit in ("day", "days"):
+        delta = timedelta(days=amount)
+    elif unit in ("week", "weeks"):
+        delta = timedelta(weeks=amount)
+    else:
+        raise argparse.ArgumentTypeError(f"Unsupported time unit: {unit}")
+
+    return datetime.now(timezone.utc) - delta
+
+
+def main() -> int:
+    parser = argparse.ArgumentParser(
+        description="Scan and clean logically deleted objects in object storage like S3.",
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+        epilog="""
+        Examples:
+        # Scan only (default)
+        %(prog)s my-bucket
+
+        # Scan with filters
+        %(prog)s my-bucket --prefix logs/ --deleted-after 2024-12-01
+
+        # Delete with confirmation prompt
+        %(prog)s my-bucket --prune
+
+        # Delete without confirmation (automation)
+        %(prog)s my-bucket --prune --noinput
+        """,
+    )
+
+    # Required
+    parser.add_argument("bucket", help="Target S3 bucket")
+
+    # Connection options
+    parser.add_argument("--prefix", help="Key prefix to scan")
+    parser.add_argument("--region", help="Region")
+    parser.add_argument("--profile", help="Profile")
+    parser.add_argument("--endpoint-url", help="Custom S3 endpoint URL")
+
+    # Filter options
+    parser.add_argument(
+        "--deleted-after",
+        type=datetime.fromisoformat,
+        help="Only process objects deleted after this date (ISO 8601 format, e.g. 2024-12-01:00:00:00) Cannot be used with `--deleted-since.`",
+    )
+
+    parser.add_argument(
+        "--deleted-since",
+        type=parse_since_cutoff,
+        metavar="DURATION",
+        help="Time duration ago (e.g. '3 days', '2 weeks'). Cannot be used with `--deleted-after.`",
+    )
+
+    # Action flags
+    parser.add_argument(
+        "--info", action="store_true", help="Show bucket information (default behavior)"
+    )
+    parser.add_argument("--scan", action="store_true", help="Scan only")
+
+    parser.add_argument(
+        "--prune",
```

Comment:
```suggestion
    parser.add_argument(
        "--permanently-delete-versions",
```

Be scary when there is a chance to destroy data. Should be very very explicit what it is going to do.